code/modules/mob/living/carbon/human/species_types/mothmen.dm

```diff
@@ -54,3 +62,136 @@
+		var/obj/structure/moth_cocoon/C = new(get_turf(H))
+		H.forceMove(C)
+		H.Sleeping(20, FALSE)
+		C.done_regenerating = TRUE
+		H.apply_status_effect(STATUS_EFFECT_COCOONED)
+		H.log_message("has finished weaving a cocoon.", LOG_GAME)
+		addtimer(CALLBACK(src, .proc/emerge, C), COCOON_EMERGE_DELAY, TIMER_UNIQUE)
+	else
+		to_chat(H, "<span class='warning'>You need to hold still in order to weave a cocoon!</span>")
+
+/datum/action/innate/cocoon/IsAvailable()
+	if(..())
+		var/mob/living/carbon/human/H = owner
+		if(HAS_TRAIT(H, TRAIT_MOTH_BURNT))
+			return TRUE
+		return FALSE
+
+//Removes moth from cocoon, restores burnt wings
+/datum/action/innate/cocoon/proc/emerge(obj/structure/moth_cocoon/C)
+	for(var/mob/living/carbon/human/H in C.contents)
+		if(!H.has_status_effect(STATUS_EFFECT_COCOONED))
+			return
+		SEND_SIGNAL(H, COMSIG_CLEAR_MOOD_EVENT, "burnt_wings")
+		if(ismoth(H) && HAS_TRAIT(H, TRAIT_MOTH_BURNT))
+			REMOVE_TRAIT(H, TRAIT_MOTH_BURNT, "fire")
+			var/obj/item/organ/wings/moth/W = H.getorgan(/obj/item/organ/wings/moth)
+			if(W)
+				W.flight_level = WINGS_FLIGHTLESS//The check for wings getting burned makes them cosmetic, so this allows the burned off effect to be applied again
+				if(locate(/datum/mutation/strongwings) in H.dna.mutations)
+					W.flight_level = WINGS_FLYING
+		H.dna.species.handle_mutant_bodyparts(H)
+		H.dna.species.handle_body(H)
+	C.done_regenerating = FALSE
+	qdel(C)
+
+/obj/structure/moth_cocoon
+	name = "\improper Mothperson cocoon"
+	desc = "Someone wrapped in a Mothperson cocoon. It's best to let them rest."
+	icon = 'icons/effects/effects.dmi'
+	icon_state = "cocoon_large1"
+	anchored = TRUE
+	max_integrity = 10
+	var/done_regenerating = TRUE//Determines whether or not the mothperson is done regenerating their wings with FALSE meaning they are
+
+/obj/structure/moth_cocoon/play_attack_sound(damage_amount, damage_type = BRUTE, damage_flag = 0)
+	switch(damage_type)
+		if(BRUTE)
+			playsound(src, 'sound/weapons/slash.ogg', 80, TRUE)
+		if(BURN)
+			playsound(src, 'sound/items/welder.ogg', 80, TRUE)
+
+/obj/structure/moth_cocoon/Destroy()
+	if(!done_regenerating)
+		visible_message("<span class='danger'>[src] splits open from within!</span>")
+	else
+		visible_message("<span class='danger'>[src] is torn open, harming the Mothperson within!</span>")
+		for(var/mob/living/carbon/human/H in contents)
+			if(H.has_status_effect(STATUS_EFFECT_COCOONED))
+				H.adjustBruteLoss(COCOON_HARM_AMOUNT, FALSE)
+				H.SetSleeping(0, FALSE)
+	for(var/mob/living/carbon/human/H in contents)
+		H.remove_status_effect(STATUS_EFFECT_COCOONED)
+		H.dna.species.handle_mutant_bodyparts(H)
+		H.dna.species.handle_body(H)
+		H.forceMove(loc)
+		if(!done_regenerating)
+			visible_message("<span class='notice'>[H]'s wings unfold, looking good as new!</span>", "<span class='notice'>Your wings unfold with new vigor!.</span>")
```

Comment:
This is also not a biggie. We can save on type-checking loops here.
Might have to manually punch this in, web-edit fucked the formatting. 
```suggestion
	if(!done_regenerating)
		visible_message("<span class='danger'>[src] splits open from within!</span>")
	else
		visible_message("<span class='danger'>[src] is torn open, harming the Mothperson within!</span>")
	for(var/mob/living/carbon/human/H in contents)
	    if(!done_regenerating)
			visible_message("<span class='notice'>[H]'s wings unfold, looking good as new!</span>", "<span class='notice'>Your wings unfold with new vigor!.</span>")
		else if(H.has_status_effect(STATUS_EFFECT_COCOONED))
				H.adjustBruteLoss(COCOON_HARM_AMOUNT, FALSE)
				H.SetSleeping(0, FALSE)
		H.remove_status_effect(STATUS_EFFECT_COCOONED)
		H.dna.species.handle_mutant_bodyparts(H)
		H.dna.species.handle_body(H)
		H.forceMove(loc)
		    
```